File: demo/java/custom-java/src/main/java/com/baidu/bce/utils/L.java
package com.baidu.bce.utils;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;

public class L {
    private L() {
    }

    private static BufferedWriter bufferedWriter;
    private static SimpleDateFormat sdf;

    static {
        String filePath = "output/custom-java.txt"; // 文件路径
        sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
        try {
            FileWriter fileWriter = new FileWriter(filePath, false);
            bufferedWriter = new BufferedWriter(fileWriter);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void info(String s) {
        text("[info] " + s);
    }

    public static void debug(String s) {
        text("[debug] " + s);
    }

    public static void warn(String s) {
        text("[warn] " + s);
    }

    public static void error(String s) {
        text("[error] " + s);
    }

    private static void text(String s) {
        s = "[" + sdf.format(new Date(System.currentTimeMillis())) + "] " + s;
        System.err.println(s);
        if (bufferedWriter != null) {
            try {
                bufferedWriter.write(s);
                bufferedWriter.newLine();
                bufferedWriter.flush();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: demo/cxx/custom-cxx/src/utils/l.cpp
#include "l.h"

std::ofstream L::bufferedWriter;
std::time_t L::currentTime;
std::tm* L::localTime;
char L::timeBuffer[80];

const std::string L::filePath = "output/custom-cpp.txt";
const std::string L::logInfo = "[info] ";
const std::string L::logDebug = "[debug] ";
const std::string L::logWarn = "[warn] ";
const std::string L::logError = "[error] ";

L::L() {}

void L::text(const std::string& s) {
    currentTime = std::time(nullptr);
    localTime = std::localtime(&currentTime);
    std::strftime(timeBuffer, sizeof(timeBuffer), "%Y-%m-%d %H:%M:%S.000", localTime);

    std::cerr << "[" << timeBuffer << "] " << s << std::endl;

    if (bufferedWriter.is_open()) {
        bufferedWriter << "[" << timeBuffer << "] " << s << std::endl;
        bufferedWriter.flush();
    }
}

void L::info(const std::string& s) {
    text(logInfo + s);
}

void L::debug(const std::string& s) {
    text(logDebug + s);
}

void L::warn(const std::string& s) {
    text(logWarn + s);
}

void L::error(const std::string& s) {
    text(logError + s);
}

